File: ui/lib/proof-verifier.ts
import { createHash, createPublicKey, verify as verifySignature } from "node:crypto";
import type { Hex, PublicClient } from "viem";

import { SYNOD_REGISTRY_ABI } from "./registry-abi";
import type { OnchainSettlement } from "./registry";
import type { ProofVerificationView, ProofVoteView } from "./types";

const ED25519_SPKI_PREFIX = Buffer.from("302a300506032b6570032100", "hex");
const SUPPORTED_PROTOCOL_VERSIONS = [1, 2];
const PROTOCOL_VERSION_LATEST = 2;

function canonicalJson(value: unknown): string {
  if (value === null || typeof value !== "object") return JSON.stringify(value);
  if (Array.isArray(value)) return `[${value.map(canonicalJson).join(",")}]`;
  const entries = Object.entries(value as Record<string, unknown>)
    .filter(([, v]) => v !== undefined)
    .sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0));
  return `{${entries
    .map(([k, v]) => `${JSON.stringify(k)}:${canonicalJson(v)}`)
    .join(",")}}`;
}

function sha256HexText(text: string): string {
  return createHash("sha256").update(text, "utf8").digest("hex");
}

function sha256HexCanonical(value: unknown): string {
  return createHash("sha256").update(canonicalJson(value), "utf8").digest("hex");
}

function fromHex(hex: string): Buffer {
  const stripped = hex.startsWith("0x") ? hex.slice(2) : hex;
  return Buffer.from(stripped, "hex");
}

function bytes32Hex(hex: string): Hex {
  const stripped = hex.startsWith("0x") ? hex.slice(2) : hex;
  return `0x${stripped}` as Hex;
}

function isHex32(value: unknown): value is string {
  return typeof value === "string" && /^[0-9a-fA-F]{64}$/.test(value);
}

function isSignatureHex(value: unknown): value is string {
  return typeof value === "string" && /^[0-9a-fA-F]{128}$/.test(value);
}

function round4(n: number): number {
  return Number(n.toFixed(4));
}

/**
 * Mirror of Python's protocol.canonical_confidence().
 *
 * Python's `json.dumps(1.0)` emits `"1.0"` while `JSON.stringify(1)` emits
 * `"1"`. Without normalization, a settler that returns `confidence == 1.0`
 * (e.g. Gemini at high certainty) signs over a payload no JS verifier can
 * reconstruct byte-identically. Clamp to [0, 0.9999] after 4-decimal
 * rounding so the value is never a whole-integer float.
 */
function canonicalConfidence(n: number): number {
  const clamped = Math.max(0, Math.min(1, Number(n)));
  const rounded = Number(clamped.toFixed(4));
  return Math.min(rounded, 0.9999);
}

function verifyEd25519(pubkeyHex: string, payload: Buffer, signatureHex: string): boolean {
  try {
    const pubkey = fromHex(pubkeyHex);
    if (pubkey.length !== 32) return false;
    const signature = fromHex(signatureHex);
    if (signature.length !== 64) return false;
    const key = createPublicKey({
      key: Buffer.concat([ED25519_SPKI_PREFIX, pubkey]),
      format: "der",
      type: "spki",
    });
    return verifySignature(null, payload, key, signature);
  } catch {
    return false;
  }
}

interface ProofQuestion {
  kind: number;
  question_id: string;
  prompt: string;
  outcomes: number[];
  deadline: number;
}

interface ProofVote {
  kind: number;
  protocol_version: number;
  question_id: string;
  prompt_hash: string;
  outcomes_hash: string;
  deadline: number;
  settler_pubkey: string;
  model_tag: string;
  outcome: number;
  confidence: number;
  timestamp: number;
  signature: string;
  // v2+: reasoning is bound into the signature via reasoning_hash.
  reasoning_hash?: string;
  // Optional plaintext reasoning carried alongside the signed vote.
  reasoning?: string;
}

interface ProofPayload {
  protocol_version?: number;
  question?: ProofQuestion;
  votes?: ProofVote[];
}

function signingPayload(vote: ProofVote): Buffer {
  const body: Record<string, unknown> = {
    kind: vote.kind,
    protocol_version: vote.protocol_version,
    question_id: vote.question_id,
    prompt_hash: vote.prompt_hash,
    outcomes_hash: vote.outcomes_hash,
    deadline: vote.deadline,
    settler_pubkey: vote.settler_pubkey.toLowerCase(),
    model_tag: vote.model_tag,
    outcome: vote.outcome,
    confidence: canonicalConfidence(Number(vote.confidence)),
    timestamp: vote.timestamp,
  };
  if (vote.protocol_version >= 2) {
    body.reasoning_hash = vote.reasoning_hash ?? "";
  }
  return Buffer.from(canonicalJson(body), "utf8");
}

export async function verifySettlementProof(
  client: PublicClient,
  registryAddress: Hex,
  settlement: OnchainSettlement
): Promise<ProofVerificationView> {
  const errors: string[] = [];
  const votesView: ProofVoteView[] = [];

  let parsed: ProofPayload;
  try {
    const text = fromHex(settlement.signedVotesPayload).toString("utf8");
    parsed = JSON.parse(text) as ProofPayload;
  } catch {
    return { status: "invalid", errors: ["signedVotesPayload is not valid JSON"], votes: [] };
  }

  if (
    typeof parsed.protocol_version !== "number" ||
    !SUPPORTED_PROTOCOL_VERSIONS.includes(parsed.protocol_version)
  ) {
    errors.push(`unsupported protocol_version ${String(parsed.protocol_version)}`);
  }
  if (!parsed.question) errors.push("proof payload is missing question");
  if (!Array.isArray(parsed.votes) || parsed.votes.length === 0) {
    errors.push("proof payload has no votes");
  }

  const question = parsed.question;
  const votes = Array.isArray(parsed.votes) ? parsed.votes : [];
  const settlementQuestionId = settlement.questionId.replace(/^0x/, "").toLowerCase();

  let promptHash = "";
  let outcomesHash = "";
  if (question) {
    if (question.question_id.toLowerCase() !== settlementQuestionId) {
      errors.push("question_id does not match on-chain settlement");
    }
    promptHash = sha256HexText(question.prompt);
    outcomesHash = sha256HexCanonical(question.outcomes);
  }

  const seen = new Set<string>();
  const outcomeCounts = new Map<number, number>();
  const outcomeWeights = new Map<number, number>();

  for (const vote of votes) {
    const pubkey = typeof vote.settler_pubkey === "string" ? vote.settler_pubkey.toLowerCase() : "";
    const row: ProofVoteView = {
      pubkey,
      modelTag: vote.model_tag,
      outcome: Number(vote.outcome),
      confidence: Number(vote.confidence),
      registered: false,
      signatureValid: false,
    };

    if (!isHex32(pubkey)) {
      errors.push("vote has invalid settler_pubkey");
      votesView.push(row);
      continue;
    }
    if (seen.has(pubkey)) {
      errors.push(`duplicate vote from ${pubkey.slice(0, 16)}`);
      votesView.push(row);
      continue;
    }
    seen.add(pubkey);

    try {
      row.registered = Boolean(
        await client.readContract({
          address: registryAddress,
          abi: SYNOD_REGISTRY_ABI,
          functionName: "registeredAxlPubKeys",
          args: [bytes32Hex(pubkey)],
        })
      );
    } catch {
      errors.push(`could not check registry membership for ${pubkey.slice(0, 16)}`);
    }
    if (!row.registered) errors.push(`unregistered AXL pubkey ${pubkey.slice(0, 16)}`);

    if (!isSignatureHex(vote.signature)) {
      errors.push(`invalid signature encoding for ${pubkey.slice(0, 16)}`);
    } else {
      row.signatureValid = verifyEd25519(pubkey, signingPayload(vote), vote.signature);
      if (!row.signatureValid) errors.push(`bad signature for ${pubkey.slice(0, 16)}`);
    }

    if (!SUPPORTED_PROTOCOL_VERSIONS.includes(vote.protocol_version)) {
      errors.push(`bad vote protocol for ${pubkey.slice(0, 16)}`);
    }
    if (vote.protocol_version >= 2) {
      // v2+: reasoning text must hash to the signed reasoning_hash field.
      const expected = sha256HexText(vote.reasoning ?? "");
      if (typeof vote.reasoning_hash !== "string" || vote.reasoning_hash.length !== 64) {
        errors.push(`missing or malformed reasoning_hash for ${pubkey.slice(0, 16)}`);
      } else if (vote.reasoning_hash !== expected) {
        errors.push(`reasoning tampering detected for ${pubkey.slice(0, 16)}`);
      }
    }
    if (vote.question_id?.toLowerCase() !== settlementQuestionId) errors.push(`vote question mismatch for ${pubkey.slice(0, 16)}`);
    if (question && vote.deadline !== question.deadline) errors.push(`vote deadline mismatch for ${pubkey.slice(0, 16)}`);
    if (question && vote.prompt_hash !== promptHash) errors.push(`vote prompt hash mismatch for ${pubkey.slice(0, 16)}`);
    if (question && vote.outcomes_hash !== outcomesHash) errors.push(`vote outcomes hash mismatch for ${pubkey.slice(0, 16)}`);
    if (question && !question.outcomes.includes(Number(vote.outcome))) errors.push(`vote outcome invalid for ${pubkey.slice(0, 16)}`);
    if (!Number.isFinite(Number(vote.confidence)) || Number(vote.confidence) < 0 || Number(vote.confidence) > 1) {
      errors.push(`vote confidence invalid for ${pubkey.slice(0, 16)}`);
    }

    votesView.push(row);
    if (row.registered && row.signatureValid) {
      const outcome = Number(vote.outcome);
      const confidence = round4(Number(vote.confidence));
      outcomeCounts.set(outcome, (outcomeCounts.get(outcome) ?? 0) + 1);
      outcomeWeights.set(outcome, (outcomeWeights.get(outcome) ?? 0) + confidence);
    }
  }

  const expectedOutcome = Number(settlement.outcome);
  const winnerVotes = outcomeCounts.get(expectedOutcome) ?? 0;
  const weightedScoreScaled = Math.round((outcomeWeights.get(expectedOutcome) ?? 0) * 1_000_000);
  if (winnerVotes < Number(settlement.quorumSize)) {
    errors.push("winning outcome does not meet on-chain quorumSize");
  }
  if (weightedScoreScaled !== Number(settlement.weightedScoreScaled)) {
    errors.push("weighted score does not match signed votes");
  }

  return {
    status: errors.length === 0 ? "verified" : "invalid",
    errors,
    questionId: settlementQuestionId,
    quorumSize: Number(settlement.quorumSize),
    winnerVotes,
    weightedScoreScaled,
    votes: votesView,
  };
}

File: settler/synod_settler/proof_verifier.py
"""Independent verifier for Synod signed-vote settlement proofs.

The verifier is deliberately shared by tests and the CLI. It does not trust
agent logs or UI state: it parses the stored proof payload, verifies ed25519
signatures against AXL pubkeys, checks registry membership via a caller-supplied
callback, and recomputes the on-chain quorum and weighted score.
"""

from __future__ import annotations

import json
import math
from collections.abc import Callable
from dataclasses import dataclass, field
from typing import Any

from .identity import verify_signature
from .protocol import (
    MessageKind,
    PROTOCOL_VERSION,
    reasoning_sha256_hex,
    QuestionAnnouncement,
    SettlementVote,
    canonical_sha256_hex,
    text_sha256_hex,
    validate_hex32,
)


SCALE = 1_000_000


@dataclass(frozen=True)
class SettlementSnapshot:
    """On-chain settlement fields needed to verify a proof payload."""

    question_id: str
    outcome: int
    quorum_size: int
    weighted_score_scaled: int

    @classmethod
    def new(
        cls,
        *,
        question_id: str,
        outcome: int,
        quorum_size: int,
        weighted_score_scaled: int,
    ) -> SettlementSnapshot:
        return cls(
            question_id=validate_hex32(question_id.removeprefix("0x"), field="question_id"),
            outcome=int(outcome),
            quorum_size=int(quorum_size),
            weighted_score_scaled=int(weighted_score_scaled),
        )


@dataclass
class VerifiedVote:
    pubkey: str
    model_tag: str
    outcome: int
    confidence: float
    registered: bool = False
    signature_valid: bool = False
    valid: bool = False
    errors: list[str] = field(default_factory=list)


@dataclass
class ProofVerification:
    status: str
    settlement: SettlementSnapshot
    winner_votes: int
    weighted_score_scaled: int
    votes: list[VerifiedVote]
    errors: list[str]

    @property
    def verified(self) -> bool:
        return self.status == "verified"

    def to_dict(self) -> dict[str, Any]:
        return {
            "status": self.status,
            "question_id": self.settlement.question_id,
            "outcome": self.settlement.outcome,
            "quorum_size": self.settlement.quorum_size,
            "winner_votes": self.winner_votes,
            "weighted_score_scaled": self.weighted_score_scaled,
            "expected_weighted_score_scaled": self.settlement.weighted_score_scaled,
            "errors": list(self.errors),
            "votes": [
                {
                    "pubkey": v.pubkey,
                    "model_tag": v.model_tag,
                    "outcome": v.outcome,
                    "confidence": v.confidence,
                    "registered": v.registered,
                    "signature_valid": v.signature_valid,
                    "valid": v.valid,
                    "errors": list(v.errors),
                }
                for v in self.votes
            ],
        }


def _is_signature_hex(value: Any) -> bool:
    if not isinstance(value, str) or len(value) != 128:
        return False
    try:
        raw = bytes.fromhex(value)
    except ValueError:
        return False
    return len(raw) == 64 and value.lower() == value.strip().lower()


def _parse_question(raw: Any, errors: list[str]) -> QuestionAnnouncement | None:
    if not isinstance(raw, dict):
        errors.append("proof payload is missing question")
        return None
    try:
        q = QuestionAnnouncement.new(
            question_id=str(raw["question_id"]),
            prompt=str(raw["prompt"]),
            outcomes=[int(x) for x in raw["outcomes"]],
            deadline=int(raw["deadline"]),
        )
    except (KeyError, TypeError, ValueError) as e:
        errors.append(f"invalid question: {e}")
        return None
    if int(raw.get("kind", MessageKind.QUESTION)) != int(MessageKind.QUESTION):
        errors.append("question has wrong kind")
    return q


def verify_proof_payload(
    payload_bytes: bytes,
    settlement: SettlementSnapshot,
    is_registered_pubkey: Callable[[str], bool],
) -> ProofVerification:
    """Verify a Synod proof payload against an expected settlement snapshot."""

    errors: list[str] = []
    verified_votes: list[VerifiedVote] = []

    try:
        parsed = json.loads(payload_bytes.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError):
        return ProofVerification(
            status="invalid",
            settlement=settlement,
            winner_votes=0,
            weighted_score_scaled=0,
            votes=[],
            errors=["signedVotesPayload is not valid JSON"],
        )

    if not isinstance(parsed, dict):
        errors.append("proof payload must be a JSON object")
        parsed = {}
    if parsed.get("protocol_version") not in (1, PROTOCOL_VERSION):
        errors.append(f"unsupported protocol_version {parsed.get('protocol_version')}")

    question = _parse_question(parsed.get("question"), errors)
    if question and question.question_id != settlement.question_id:
        errors.append("question_id does not match settlement")

    raw_votes = parsed.get("votes")
    if not isinstance(raw_votes, list) or not raw_votes:
        errors.append("proof payload has no votes")
        raw_votes = []

    seen: set[str] = set()
    outcome_counts: dict[int, int] = {}
    outcome_weights: dict[int, float] = {}

    for index, raw_vote in enumerate(raw_votes):
        if not isinstance(raw_vote, dict):
            errors.append(f"vote {index} is not an object")
            continue

        vote_errors: list[str] = []
        pubkey = ""
        model_tag = str(raw_vote.get("model_tag", ""))
        outcome = -1
        confidence = math.nan
        try:
            pubkey = validate_hex32(str(raw_vote.get("settler_pubkey", "")), field="settler_pubkey")
        except ValueError as e:
            vote_errors.append(str(e))

        try:
            outcome = int(raw_vote["outcome"])
            confidence = float(raw_vote["confidence"])
        except (KeyError, TypeError, ValueError) as e:
            vote_errors.append(f"invalid outcome/confidence: {e}")

        row = VerifiedVote(
            pubkey=pubkey,
            model_tag=model_tag,
            outcome=outcome,
            confidence=confidence,
        )

        if pubkey:
            if pubkey in seen:
                vote_errors.append("duplicate vote from settler")
            seen.add(pubkey)
            try:
                row.registered = bool(is_registered_pubkey(pubkey))
            except Exception as e:  # pragma: no cover - defensive around RPC failures
                vote_errors.append(f"registry membership check failed: {e}")
            if not row.registered:
                vote_errors.append("settler pubkey is not registered")

        signature = str(raw_vote.get("signature", ""))
        if not _is_signature_hex(signature):
            vote_errors.append("signature must be 64-byte hex")

        try:
            protocol_version = int(raw_vote["protocol_version"])
            kind = int(raw_vote["kind"])
            deadline = int(raw_vote["deadline"])
            timestamp = int(raw_vote["timestamp"])
            prompt_hash = str(raw_vote["prompt_hash"])
            outcomes_hash = str(raw_vote["outcomes_hash"])
            question_id = validate_hex32(str(raw_vote["question_id"]), field="question_id")
        except (KeyError, TypeError, ValueError) as e:
            vote_errors.append(f"malformed signed vote fields: {e}")
        else:
            if kind != int(MessageKind.VOTE):
                vote_errors.append("vote has wrong kind")
            if protocol_version not in (1, PROTOCOL_VERSION):
                vote_errors.append("vote has unsupported protocol_version")
            if question_id != settlement.question_id:
                vote_errors.append("vote question_id does not match settlement")
            if question:
                if deadline != question.deadline:
                    vote_errors.append("vote deadline does not match question")
                if prompt_hash != text_sha256_hex(question.prompt):
                    vote_errors.append("vote prompt_hash does not match question")
                if outcomes_hash != canonical_sha256_hex(question.outcomes):
                    vote_errors.append("vote outcomes_hash does not match question")
                if outcome not in question.outcomes:
                    vote_errors.append("vote outcome is not in question outcomes")
            if timestamp > deadline:
                vote_errors.append("vote timestamp is after deadline")
            if not math.isfinite(confidence) or confidence < 0.0 or confidence > 1.0:
                vote_errors.append("vote confidence is outside [0, 1]")

            # v2+: reasoning_hash must be present and bind the displayed reasoning
            reasoning_hash = ""
            if protocol_version >= 2:
                rh = raw_vote.get("reasoning_hash")
                reasoning_text = str(raw_vote.get("reasoning", ""))
                expected = reasoning_sha256_hex(reasoning_text)
                if not isinstance(rh, str) or len(rh) != 64:
                    vote_errors.append("missing or malformed reasoning_hash")
                elif rh != expected:
                    vote_errors.append("reasoning text does not match signed reasoning_hash")
                reasoning_hash = rh if isinstance(rh, str) else ""

            if pubkey and _is_signature_hex(signature):
                vote_for_signing = SettlementVote(
                    kind=kind,
                    protocol_version=protocol_version,
                    question_id=question_id,
                    prompt_hash=prompt_hash,
                    outcomes_hash=outcomes_hash,
                    deadline=deadline,
                    settler_pubkey=pubkey,
                    model_tag=model_tag,
                    outcome=outcome,
                    confidence=confidence,
                    timestamp=timestamp,
                    reasoning_hash=reasoning_hash,
                )
                row.signature_valid = verify_signature(
                    pubkey,
                    vote_for_signing.signing_payload(),
                    signature,
                )
                if not row.signature_valid:
                    vote_errors.append("ed25519 signature is invalid")

        row.errors = vote_errors
        row.valid = not vote_errors
        verified_votes.append(row)
        errors.extend(f"{pubkey[:16] or f'vote {index}'}: {e}" for e in vote_errors)

        if row.valid:
            outcome_counts[outcome] = outcome_counts.get(outcome, 0) + 1
            outcome_weights[outcome] = outcome_weights.get(outcome, 0.0) + round(confidence, 4)

    winner_votes = outcome_counts.get(settlement.outcome, 0)
    weighted_score_scaled = int(round(outcome_weights.get(settlement.outcome, 0.0) * SCALE))
    if winner_votes < settlement.quorum_size:
        errors.append("winning outcome does not meet settlement quorum_size")
    if weighted_score_scaled != settlement.weighted_score_scaled:
        errors.append("weighted score does not match settlement")

    return ProofVerification(
        status="verified" if not errors else "invalid",
        settlement=settlement,
        winner_votes=winner_votes,
        weighted_score_scaled=weighted_score_scaled,
        votes=verified_votes,
        errors=errors,
    )
